Small fixes from book club.

jdunkerley opened the pull request:
### Pull Request Description

- Add the missing dropdowns for `Locale` and `Encoding`.
- Correct a few mismatched type signatures.
- Adjust `order_by` calls with a single `Sort_Column` to call in a Vector.
- Adjust parameter names for `transpose`.
- Fix for the table viz: escape HTML and `suppressFieldDotNotation`.
- Use `Filter_Condition.Equal True` for the default filter.
- Adjust `Data.fetch` to return the response on success when parse fails. Rename `parse` to `try_auto_parse`.
- Add various aliases for methods.
- Add tests for `Table.set` when using a `Vector`, `Range` or `Date_Range`.
- Add check for mismatched length on `Table.set`.

![image](https://github.com/enso-org/enso/assets/4699705/23ea0ba3-2b05-4af8-afd9-f35b55446c24)

![image](https://github.com/enso-org/enso/assets/4699705/8b0253e6-e9e8-490a-9607-0da51ab5a215)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

jdunkerley committed:
Fix type signatures again.
Add dropdown for Delimited.

jdunkerley committed:
Add a load of missing dropdowns.

jdunkerley committed:
Missing import.

jdunkerley committed:
Remove non-vector wrapped use of Sort_Column.

jdunkerley committed:
Rename `transpose` parameter:
- id_fields => key_columns
- name_field => attribute_column_name
- value_field => value_column_name

jdunkerley committed:
Fix for `table.js`:
- Escape HTML.
- Suppress dot notation in fields.

jdunkerley committed:
Use `Equal True` for default filtering.

jdunkerley committed:
Adjust fetch to return response on error.
Rename `parse` to `try_auto_parse`.

jdunkerley committed:
Add aliases for `fill_nothing` and `fill_empty`.

jdunkerley committed:
Add aliases for `sort` and `order_by`.

jdunkerley committed:
Add tests for Table.set from Vector and Range.
Check length for in-memory set.

GregoryTravis commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso:
```diff
@@ -1142,15 +1144,15 @@ round_min_long = -99999999999999
 ## PRIVATE
    Restrict rounding decimal_places parameter.
-check_decimal_places : Integer -> Function
+check_decimal_places : Integer -> Function -> Any ! Illegal_Argument
```
```suggestion
check_decimal_places : Integer -> Any -> Any ! Illegal_Argument
```

Because action is Any and is returned unless there is an error.

GregoryTravis commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -1373,12 +1373,12 @@ type Table
            `Duplicate_Output_Column_Names` is reported according to the
            `on_problems` setting.
-    @id_fields Widget_Helpers.make_column_name_vector_selector
+    @key_columns Widget_Helpers.make_column_name_vector_selector
```
I think this one could really use a small ascii-table example.

jdunkerley committed:
Update distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso

Co-authored-by: GregoryTravis <greg.m.travis@gmail.com>

farmaazon reviewed (approved):
Approving changes in table.js.

jdunkerley committed:
Examples for CrossTab and Transpose.

radeusgd commented on app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:
```diff
@@ -99,7 +99,13 @@ class TableVisualization extends Visualization {
                 return '<span style="color:grey; font-style: italic;">Empty</span>'
             }
-            return params.value.toString()
+            return params.value
+                .toString()
+                .replace(
+                    /[&<"'>]/g,
+                    m =>
+                        ({ '&': '&amp;', '<': '&lt;', '"': '&quot;', "'": '&#39;', '>': '&gt;' }[m])
+                )
```
```suggestion
            const mapping = { '&': '&amp;', '<': '&lt;', '"': '&quot;', "'": '&#39;', '>': '&gt;' };
            return params.value
                .toString()
                .replace(
                    /[&<"'>]/g,
                    m => mapping[m]
                )
```

I'm not sure how good the JS optimizer is, but it feels like it will have an easier job if we initialize this map only once and not on each callback-call (which can be a lot if some huge HTML is being escaped).

Also, I think we should move this into a separate function:

```
function escapeHTML(code) { ... }
```

It will make the code more readable.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text.enso:
```diff
@@ -55,15 +55,15 @@ type Text
 
     ## ADVANCED
-       PRIVATE
-       UNSTABLE
        Unifies the case of all letters in the text, generating a key which can be
        used to perform case-insensitive comparisons.
+    @locale Locale.default_widget
     to_case_insensitive_key : Locale -> Text
     to_case_insensitive_key self locale=Locale.default =
```
I'm not completely sure if we should un`PRIVATE` this - as it is a bit of an implementation detail. But I appreciate it can be useful to users too.

radeusgd commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -1373,12 +1373,30 @@ type Table
            `on_problems` setting.
-    @id_fields Widget_Helpers.make_column_name_vector_selector
+
+       ? Example Transpose Operation
+
+         Input Table `table`:
+
+            A | Name    | Country
+           ---|---------|---------
+            A | Example | France
+            B | Another | Germany
+
+         Result `table.transpose ['A'] 'Attribute' 'Value'`:
+
+            A | Attribute | Value
+           ---|-----------|---------
+            A | Name      | Example
+            A | Country   | France
+            B | Name      | Another
+            B | Country   | Germany
```
THIS IS SO GREAT!

I used to not know what transpose does and I would have really appreciate this example back in the day.

I'd just suggest to maybe rename the column `A` as it clashes with the value `A` and while the context is rather clear, it can lead to some ambiguity. Maybe let's call the column `Id` or `X`?

radeusgd commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -1414,6 +1432,21 @@ type Table
            - If there are more than 10 issues with a single column,
              an `Additional_Warnings`.
+
+       ? Example Cross Tab Operation
+
+         Input Table `table`:
+
+            A | B       | C
+           ---|---------|---------
+            A | Name    | Example
+            A | Country | France
+
+         Result `table.cross_tab ['A'] 'B' (Aggregate_Column.First 'C')`:
+
+            A | Name    | Country
+           ---|---------|---------
+            A | Example | France
```
Btw. this leads me to thinking.

If I want to reconstruct the table from an earlier `transpose` using `cross_tab`, I'll do sth like you did here.

Shall we have something like `Aggregate_Column.Only_Row` which will act like `First` but if there is more than 1 record in each group it would error? This way I can make sure that the reverse-transpose worked correctly and verify if there were no data corruptions along the way.

Not sure how needed that is, but I thought it might be useful.

radeusgd commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -1259,18 +1259,20 @@ type Table
-        check_add_mode.if_not_error <|
-            Table.Value (self.java_table.addOrReplaceColumn renamed.java_column)
+
+        if resolved.length != self.row_count then Error.throw (Illegal_Argument.Error "The length of the column does not match the number of rows in the table.") else
```
Shall we use the `Row_Count_Mismatch` from `Table.zip` instead?

radeusgd reviewed (approved):
Looks all good, just a few comments.

jdunkerley committed:
Fix CodeGen issue caused by change to Filter_Condition.
Make escapeHTML a function.
Hide to_case_insensitive_key.
Use Row_Count_Mismatch.

jdunkerley committed:
Missing commit.

jdunkerley committed:
Failed test.
